File: src/src/main/resources/mmlspark/RankingEvaluator.py
# Copyright (C) Microsoft Corporation. All rights reserved.
# Licensed under the MIT License. See LICENSE in project root for information.


import sys
if sys.version >= '3':
    basestring = str

from pyspark.ml.param.shared import *
from pyspark import keyword_only
from pyspark.ml.util import JavaMLReadable, JavaMLWritable
from pyspark.ml.evaluation import JavaEvaluator
from pyspark.ml.common import inherit_doc
from mmlspark.Utils import *

@inherit_doc
class RankingEvaluator(ComplexParamsMixin, JavaMLReadable, JavaMLWritable, JavaEvaluator):
    """


    Args:

        itemCol (str): Column of items
        k (int): number of items (default: 10)
        labelCol (str): label column name (default: label)
        metricName (str): metric name in evaluation (ndcgAt|map|precisionAtk|recallAtK|diversityAtK|maxDiversity|mrr|fcp) (default: ndcgAt)
        nItems (long): number of items (default: -1)
        predictionCol (str): prediction column name (default: prediction)
        ratingCol (str): Column of ratings
        userCol (str): Column of users
    """

    @keyword_only
    def __init__(self, itemCol=None, k=10, labelCol="label", metricName="ndcgAt", nItems=-1, predictionCol="prediction", ratingCol=None, userCol=None):
        super(RankingEvaluator, self).__init__()
        self._java_obj = self._new_java_obj("com.microsoft.ml.spark.RankingEvaluator")
        self.itemCol = Param(self, "itemCol", "itemCol: Column of items")
        self.k = Param(self, "k", "k: number of items (default: 10)")
        self._setDefault(k=10)
        self.labelCol = Param(self, "labelCol", "labelCol: label column name (default: label)")
        self._setDefault(labelCol="label")
        self.metricName = Param(self, "metricName", "metricName: metric name in evaluation (ndcgAt|map|precisionAtk|recallAtK|diversityAtK|maxDiversity|mrr|fcp) (default: ndcgAt)")
        self._setDefault(metricName="ndcgAt")
        self.nItems = Param(self, "nItems", "nItems: number of items (default: -1)")
        self._setDefault(nItems=-1)
        self.predictionCol = Param(self, "predictionCol", "predictionCol: prediction column name (default: prediction)")
        self._setDefault(predictionCol="prediction")
        self.ratingCol = Param(self, "ratingCol", "ratingCol: Column of ratings")
        self.userCol = Param(self, "userCol", "userCol: Column of users")
        if hasattr(self, "_input_kwargs"):
            kwargs = self._input_kwargs
        else:
            kwargs = self.__init__._input_kwargs
        self.setParams(**kwargs)

    @keyword_only
    def setParams(self, itemCol=None, k=10, labelCol="label", metricName="ndcgAt", nItems=-1, predictionCol="prediction", ratingCol=None, userCol=None):
        """
        Set the (keyword only) parameters

        Args:

            itemCol (str): Column of items
            k (int): number of items (default: 10)
            labelCol (str): label column name (default: label)
            metricName (str): metric name in evaluation (ndcgAt|map|precisionAtk|recallAtK|diversityAtK|maxDiversity|mrr|fcp) (default: ndcgAt)
            nItems (long): number of items (default: -1)
            predictionCol (str): prediction column name (default: prediction)
            ratingCol (str): Column of ratings
            userCol (str): Column of users
        """
        if hasattr(self, "_input_kwargs"):
            kwargs = self._input_kwargs
        else:
            kwargs = self.__init__._input_kwargs
        return self._set(**kwargs)

    def setItemCol(self, value):
        """

        Args:

            itemCol (str): Column of items

        """
        self._set(itemCol=value)
        return self


    def getItemCol(self):
        """

        Returns:

            str: Column of items
        """
        return self.getOrDefault(self.itemCol)


    def setK(self, value):
        """

        Args:

            k (int): number of items (default: 10)

        """
        self._set(k=value)
        return self


    def getK(self):
        """

        Returns:

            int: number of items (default: 10)
        """
        return self.getOrDefault(self.k)


    def setLabelCol(self, value):
        """

        Args:

            labelCol (str): label column name (default: label)

        """
        self._set(labelCol=value)
        return self


    def getLabelCol(self):
        """

        Returns:

            str: label column name (default: label)
        """
        return self.getOrDefault(self.labelCol)


    def setMetricName(self, value):
        """

        Args:

            metricName (str): metric name in evaluation (ndcgAt|map|precisionAtk|recallAtK|diversityAtK|maxDiversity|mrr|fcp) (default: ndcgAt)

        """
        self._set(metricName=value)
        return self


    def getMetricName(self):
        """

        Returns:

            str: metric name in evaluation (ndcgAt|map|precisionAtk|recallAtK|diversityAtK|maxDiversity|mrr|fcp) (default: ndcgAt)
        """
        return self.getOrDefault(self.metricName)


    def setNItems(self, value):
        """

        Args:

            nItems (long): number of items (default: -1)

        """
        self._set(nItems=value)
        return self


    def getNItems(self):
        """

        Returns:

            long: number of items (default: -1)
        """
        return self.getOrDefault(self.nItems)


    def setPredictionCol(self, value):
        """

        Args:

            predictionCol (str): prediction column name (default: prediction)

        """
        self._set(predictionCol=value)
        return self


    def getPredictionCol(self):
        """

        Returns:

            str: prediction column name (default: prediction)
        """
        return self.getOrDefault(self.predictionCol)


    def setRatingCol(self, value):
        """

        Args:

            ratingCol (str): Column of ratings

        """
        self._set(ratingCol=value)
        return self


    def getRatingCol(self):
        """

        Returns:

            str: Column of ratings
        """
        return self.getOrDefault(self.ratingCol)


    def setUserCol(self, value):
        """

        Args:

            userCol (str): Column of users

        """
        self._set(userCol=value)
        return self


    def getUserCol(self):
        """

        Returns:

            str: Column of users
        """
        return self.getOrDefault(self.userCol)



    @classmethod
    def read(cls):
        """ Returns an MLReader instance for this class. """
        return JavaMMLReader(cls)

    @staticmethod
    def getJavaPackage():
        """ Returns package name String. """
        return "com.microsoft.ml.spark.RankingEvaluator"

    @staticmethod
    def _from_java(java_stage):
        module_name=RankingEvaluator.__module__
        module_name=module_name.rsplit(".", 1)[0] + ".RankingEvaluator"
        return from_java(java_stage, module_name)

File: dotnetSanityCheck/dotnetClasses/RankingEvaluator.cs
// Copyright (C) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License. See LICENSE in project root for information.


using System;
using System.Collections.Generic;
using System.Linq;
using System.Reflection;
using Microsoft.Spark.ML.Feature.Param;
using Microsoft.Spark.Interop;
using Microsoft.Spark.Interop.Ipc;
using Microsoft.Spark.ML.Feature;
using Microsoft.Spark.Sql;
using Microsoft.Spark.Sql.Types;

using SynapseML.Dotnet.Utils;


namespace Synapse.ML.Recommendation
{
    /// <summary>
    /// <see cref="RankingEvaluator"/> implements RankingEvaluator
    /// </summary>
    public class RankingEvaluator : JavaEvaluator, IJavaMLWritable, IJavaMLReadable<RankingEvaluator>
    {
        private static readonly string s_className = "com.microsoft.azure.synapse.ml.recommendation.RankingEvaluator";

        /// <summary>
        /// Creates a <see cref="RankingEvaluator"/> without any parameters.
        /// </summary>
        public RankingEvaluator() : base(s_className)
        {
        }

        /// <summary>
        /// Creates a <see cref="RankingEvaluator"/> with a UID that is used to give the
        /// <see cref="RankingEvaluator"/> a unique ID.
        /// </summary>
        /// <param name="uid">An immutable unique ID for the object and its derivatives.</param>
        public RankingEvaluator(string uid) : base(s_className, uid)
        {
        }

        internal RankingEvaluator(JvmObjectReference jvmObject) : base(jvmObject)
        {
        }

        /// <summary>
        /// Sets itemCol value for <see cref="itemCol"/>
        /// </summary>
        /// <param name="itemCol">
        /// Column of items
        /// </param>
        /// <returns> New RankingEvaluator object </returns>
        public RankingEvaluator SetItemCol(string value) =>
            WrapAsRankingEvaluator(Reference.Invoke("setItemCol", (object)value));
        
        /// <summary>
        /// Sets k value for <see cref="k"/>
        /// </summary>
        /// <param name="k">
        /// number of items
        /// </param>
        /// <returns> New RankingEvaluator object </returns>
        public RankingEvaluator SetK(int value) =>
            WrapAsRankingEvaluator(Reference.Invoke("setK", (object)value));
        
        /// <summary>
        /// Sets labelCol value for <see cref="labelCol"/>
        /// </summary>
        /// <param name="labelCol">
        /// label column name
        /// </param>
        /// <returns> New RankingEvaluator object </returns>
        public RankingEvaluator SetLabelCol(string value) =>
            WrapAsRankingEvaluator(Reference.Invoke("setLabelCol", (object)value));
        
        /// <summary>
        /// Sets metricName value for <see cref="metricName"/>
        /// </summary>
        /// <param name="metricName">
        /// metric name in evaluation (ndcgAt|map|precisionAtk|recallAtK|diversityAtK|maxDiversity|mrr|fcp)
        /// </param>
        /// <returns> New RankingEvaluator object </returns>
        public RankingEvaluator SetMetricName(string value) =>
            WrapAsRankingEvaluator(Reference.Invoke("setMetricName", (object)value));
        
        /// <summary>
        /// Sets nItems value for <see cref="nItems"/>
        /// </summary>
        /// <param name="nItems">
        /// number of items
        /// </param>
        /// <returns> New RankingEvaluator object </returns>
        public RankingEvaluator SetNItems(long value) =>
            WrapAsRankingEvaluator(Reference.Invoke("setNItems", (object)value));
        
        /// <summary>
        /// Sets predictionCol value for <see cref="predictionCol"/>
        /// </summary>
        /// <param name="predictionCol">
        /// prediction column name
        /// </param>
        /// <returns> New RankingEvaluator object </returns>
        public RankingEvaluator SetPredictionCol(string value) =>
            WrapAsRankingEvaluator(Reference.Invoke("setPredictionCol", (object)value));
        
        /// <summary>
        /// Sets ratingCol value for <see cref="ratingCol"/>
        /// </summary>
        /// <param name="ratingCol">
        /// Column of ratings
        /// </param>
        /// <returns> New RankingEvaluator object </returns>
        public RankingEvaluator SetRatingCol(string value) =>
            WrapAsRankingEvaluator(Reference.Invoke("setRatingCol", (object)value));
        
        /// <summary>
        /// Sets userCol value for <see cref="userCol"/>
        /// </summary>
        /// <param name="userCol">
        /// Column of users
        /// </param>
        /// <returns> New RankingEvaluator object </returns>
        public RankingEvaluator SetUserCol(string value) =>
            WrapAsRankingEvaluator(Reference.Invoke("setUserCol", (object)value));

        
        /// <summary>
        /// Gets itemCol value for <see cref="itemCol"/>
        /// </summary>
        /// <returns>
        /// itemCol: Column of items
        /// </returns>
        public string GetItemCol() =>
            (string)Reference.Invoke("getItemCol");
        
        
        /// <summary>
        /// Gets k value for <see cref="k"/>
        /// </summary>
        /// <returns>
        /// k: number of items
        /// </returns>
        public int GetK() =>
            (int)Reference.Invoke("getK");
        
        
        /// <summary>
        /// Gets labelCol value for <see cref="labelCol"/>
        /// </summary>
        /// <returns>
        /// labelCol: label column name
        /// </returns>
        public string GetLabelCol() =>
            (string)Reference.Invoke("getLabelCol");
        
        
        /// <summary>
        /// Gets metricName value for <see cref="metricName"/>
        /// </summary>
        /// <returns>
        /// metricName: metric name in evaluation (ndcgAt|map|precisionAtk|recallAtK|diversityAtK|maxDiversity|mrr|fcp)
        /// </returns>
        public string GetMetricName() =>
            (string)Reference.Invoke("getMetricName");
        
        
        /// <summary>
        /// Gets nItems value for <see cref="nItems"/>
        /// </summary>
        /// <returns>
        /// nItems: number of items
        /// </returns>
        public long GetNItems() =>
            (long)Reference.Invoke("getNItems");
        
        
        /// <summary>
        /// Gets predictionCol value for <see cref="predictionCol"/>
        /// </summary>
        /// <returns>
        /// predictionCol: prediction column name
        /// </returns>
        public string GetPredictionCol() =>
            (string)Reference.Invoke("getPredictionCol");
        
        
        /// <summary>
        /// Gets ratingCol value for <see cref="ratingCol"/>
        /// </summary>
        /// <returns>
        /// ratingCol: Column of ratings
        /// </returns>
        public string GetRatingCol() =>
            (string)Reference.Invoke("getRatingCol");
        
        
        /// <summary>
        /// Gets userCol value for <see cref="userCol"/>
        /// </summary>
        /// <returns>
        /// userCol: Column of users
        /// </returns>
        public string GetUserCol() =>
            (string)Reference.Invoke("getUserCol");

        
        /// <summary>
        /// Loads the <see cref="RankingEvaluator"/> that was previously saved using Save(string).
        /// </summary>
        /// <param name="path">The path the previous <see cref="RankingEvaluator"/> was saved to</param>
        /// <returns>New <see cref="RankingEvaluator"/> object, loaded from path.</returns>
        public static RankingEvaluator Load(string path) => WrapAsRankingEvaluator(
            SparkEnvironment.JvmBridge.CallStaticJavaMethod(s_className, "load", path));
        
        /// <summary>
        /// Saves the object so that it can be loaded later using Load. Note that these objects
        /// can be shared with Scala by Loading or Saving in Scala.
        /// </summary>
        /// <param name="path">The path to save the object to</param>
        public void Save(string path) => Reference.Invoke("save", path);
        
        /// <returns>a <see cref="JavaMLWriter"/> instance for this ML instance.</returns>
        public JavaMLWriter Write() =>
            new JavaMLWriter((JvmObjectReference)Reference.Invoke("write"));
        
        /// <returns>an <see cref="JavaMLReader"/> instance for this ML instance.</returns>
        public JavaMLReader<RankingEvaluator> Read() =>
            new JavaMLReader<RankingEvaluator>((JvmObjectReference)Reference.Invoke("read"));

        private static RankingEvaluator WrapAsRankingEvaluator(object obj) =>
            new RankingEvaluator((JvmObjectReference)obj);

        
    }
}

        